### PR TITLE
Add support for IPv6 gateway, and use it to make examples IPv6-capable

### PR DESCRIPTION
**Pull request by progval**
Also contains a fix to #205 (`Protocol::Icmp` -> `Protocol::Icmpv6`).

Both examples were tested on my computer with both IP versions.

Instructions in the README should work, but they are a summary of a lot of trial-and-error to configure my computer so some command(s) might be missing.

### CODE REVIEW

File: examples/ping.rs [hunk 1]
```diff
@@ -15,12 +15,47 @@ use smoltcp::time::{Duration, Instant};
...
+            .unwrap();
+
+        let mut icmp_packet = $packet_type::new(icmp_payload);
+        (icmp_repr, icmp_packet)
+    }}
+}
+
+macro_rules! get_icmp_pong {
+    ( $repr_type:ident, $repr:expr, $payload:expr, $waiting_queue:expr, $remote_addr:expr,
+      $timestamp:expr, $received:expr) => {{
```
Comment (dlrobertson) [inline examples/ping.rs]: nit: whitespace.

File: examples/ping.rs [hunk 2]
```diff
@@ -15,12 +15,47 @@ use smoltcp::time::{Duration, Instant};
...
+                    Ipv6Address, Icmpv6Repr, Icmpv6Packet,
                     Ipv4Address, Icmpv4Repr, Icmpv4Packet};
 use smoltcp::iface::{NeighborCache, EthernetInterfaceBuilder};
 use smoltcp::socket::{SocketSet, IcmpSocket, IcmpSocketBuffer, IcmpPacketMetadata, IcmpEndpoint};
 use std::collections::HashMap;
 use byteorder::{ByteOrder, NetworkEndian};
 
+macro_rules! send_icmp_ping {
+    ( $repr_type:ident, $packet_type:ident, $ident:expr, $seq_no:expr, $echo_payload:expr,
+      $socket:expr, $remote_addr:expr, $device_caps:expr ) => {{
```
Comment (dlrobertson) [inline examples/ping.rs]: `device_caps` isn't used and what is the purpose of the double `{`?

File: src/iface/ethernet.rs [hunk 3]
```diff
@@ -65,6 +65,8 @@ struct InterfaceInner<'b, 'c> {
     ip_addrs:               ManagedSlice<'c, IpCidr>,
     #[cfg(feature = "proto-ipv4")]
     ipv4_gateway:           Option<Ipv4Address>,
+    #[cfg(feature = "proto-ipv6")]
+    ipv6_gateway:           Option<Ipv6Address>,
```
Comment (dlrobertson) [inline src/iface/ethernet.rs]: This works for now, but we really need to add support for NDISC Router Advertisements and Solicitations. In IPv6 this should actually be `default_routes` or something like that, where some could have been statically assigned and some could have been dynamically acquired via a Router Advertisement.

File: examples/ping.rs (hunk 2 above)
Comment (ProgVal) [reply examples/ping.rs]: I forgot to remove `device_caps` after an intermediate version.

The first `{` is part of the `macro_rules!` syntax, the second one means it's a block.

Comment (ProgVal) [reply examples/ping.rs]: I forgot to remove `device_caps` after an intermediate version.

The first `{` is part of the `macro_rules!` syntax, the second one means it's a block.

File: src/iface/ethernet.rs (hunk 3 above)
Comment (ProgVal) [reply src/iface/ethernet.rs]: Done, I added support for routes, to both IPv4 and IPv6.

File: examples/ping.rs (hunk 2 above)
Comment (dlrobertson) [reply examples/ping.rs]: Ah right, because the body is essentially a rvalue.

File: src/iface/ethernet.rs [hunk 4]
```diff
@@ -59,27 +60,25 @@ pub struct Interface<'b, 'c, DeviceT: for<'d> Device<'d>> {
 /// the `device` mutably until they're used, which makes it impossible to call other
 /// methods on the `Interface` in this time (since its `device` field is borrowed
 /// exclusively). However, it is still possible to call methods on its `inner` field.
-struct InterfaceInner<'b, 'c> {
+struct InterfaceInner<'b, 'c, 'e> {
     neighbor_cache:         NeighborCache<'b>,
     ethernet_addr:          EthernetAddress,
     ip_addrs:               ManagedSlice<'c, IpCidr>,
-    #[cfg(feature = "proto-ipv4")]
-    ipv4_gateway:           Option<Ipv4Address>,
```
Comment (dlrobertson) [inline src/iface/ethernet.rs]: I really like the idea of working on routing, but anything having to do with routes should probably be a separate PR.

**Comment by ProgVal** [comment]
I could remove the commits with routes, and put them in a new PR after
you merged this one. Would that be ok?

**Comment by dlrobertson** [comment]
> I could remove the commits with routes, and put them in a new PR

Yeah, I think that would probably be better.

**Comment by ProgVal** [comment]
done

File: examples/ping.rs [hunk 5]
```diff
@@ -52,19 +87,23 @@ fn main() {
...
     let icmp_rx_buffer = IcmpSocketBuffer::new(vec![IcmpPacketMetadata::EMPTY], vec![0; 256]);
     let icmp_tx_buffer = IcmpSocketBuffer::new(vec![IcmpPacketMetadata::EMPTY], vec![0; 256]);
     let icmp_socket = IcmpSocket::new(icmp_rx_buffer, icmp_tx_buffer);
 
     let ethernet_addr = EthernetAddress([0x02, 0x00, 0x00, 0x00, 0x00, 0x02]);
-    let ip_addr = IpCidr::new(IpAddress::from(local_addr), 24);
+    let src_ipv6 = IpAddress::v6(0xfd80, 0, 0, 0, 0, 0, 0, 1);
+    let ip_addrs = [IpCidr::new(IpAddress::v4(192, 168, 69, 1), 24),
+                    IpCidr::new(src_ipv6, 64),
+                    IpCidr::new(IpAddress::v6(0xfe80, 0, 0, 0, 0, 0, 0, 1), 64)];
```
Comment (dlrobertson) [inline examples/ping.rs]: This is the same address as `src_ipv6` right?

File: examples/httpclient.rs [hunk 6]
```diff
@@ -42,13 +42,17 @@ fn main() {
     let tcp_socket = TcpSocket::new(tcp_rx_buffer, tcp_tx_buffer);
 
     let ethernet_addr = EthernetAddress([0x02, 0x00, 0x00, 0x00, 0x00, 0x02]);
-    let ip_addrs = [IpCidr::new(IpAddress::v4(192, 168, 69, 1), 24)];
+    let ip_addrs = [IpCidr::new(IpAddress::v4(192, 168, 69, 1), 24),
+                    IpCidr::new(IpAddress::v6(0xfd80, 0, 0, 0, 0, 0, 0, 1), 64),
+                    IpCidr::new(IpAddress::v6(0xfe80, 0, 0, 0, 0, 0, 0, 1), 64)];
```
[the remark this answers is not in the document]
Comment (ProgVal) [reply examples/httpclient.rs]: No, one is link-local (fe80::), the other is not (fd80::)

File: README.md [hunk 7]
```diff
@@ -190,13 +190,19 @@ a specific user:
 sudo ip tuntap add name tap0 mode tap user $USER
 sudo ip link set tap0 up
 sudo ip addr add 192.168.69.100/24 dev tap0
+sudo ip -6 addr add fe80::100/64 dev tap0
+sudo ip -6 addr add fd80::100/64 dev tap0
```
Comment (dlrobertson) [inline README.md]: Address listed twice (for real this time :smile:)

File: README.md [hunk 8]
```diff
@@ -190,13 +190,19 @@ a specific user:
...
+sudo ip -6 route fe80::/64 dev tap0
+sudo ip -6 route fd80::/64 dev tap0
```
Comment (dlrobertson) [inline README.md]: Same. This command also fails for me. Did you mean `ip -6 route add default via ...`?

File: examples/httpclient.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (dlrobertson) [reply examples/httpclient.rs]: My bad.

File: README.md (hunk 7 above)
Comment (dlrobertson) [reply README.md]: Ugh I'm wrong again! The `fe` -> `fd` keeps confusing me lol

**Comment by dlrobertson** [comment]
Looks good once the `route` command is fixed. Thanks for working on this!

File: README.md (hunk 8 above)
Comment (ProgVal) [reply README.md]: Oh, I forgot the `add` after `route`

File: README.md (hunk 7 above)
Comment (ProgVal) [reply README.md]: Do you think I should use a different prefix to make it clearer? (eg. `fdaa::/64`)

Comment (dlrobertson) [reply README.md]: Yeah, that would be great! Could you also squash your commits?

Comment (ProgVal) [reply README.md]: Done

**Comment by dlrobertson** [comment]
@m-labs-homu r+

**Comment by m-labs-homu** [comment]
:pushpin: Commit c08b8b5 has been approved by `dlrobertson`

**Comment by m-labs-homu** [comment]
:hourglass: Testing commit c08b8b50ba93a4b3628748b4bc37753308e106cc with merge 4bf917ced35ab60ae5dd270e1582550545ec35f9...

**Comment by whitequark** [comment]
@dlrobertson Thank you for reviewing this! I am very happy with how the code turned out, and even more so that I have been freed to do something else :D

**Comment by dlrobertson** [comment]
> I am very happy with how the code turned out

@ProgVal thanks for sticking with it! Sorry for criticizing you for duplicating lines so many times :laughing:  Also could you open up an RFC issue of sorts before you create a PR for multiple routes. I think some time spent discussing a design could be useful.

**Comment by m-labs-homu** [comment]
:sunny: Test successful - [status-travis](https://travis-ci.org/m-labs/smoltcp/builds/378526318?utm_source=github_status&utm_medium=notification)
Approved by: dlrobertson
Pushing 4bf917ced35ab60ae5dd270e1582550545ec35f9 to master...
